### ISSUE TITLE
gitignore

### ISSUE THREAD

**Issue by mac2000** [trufflesecurity/trufflehog#865]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

gitignore support

## Problem to be Addressed

Scanning of filesystem may become a challenging task because there are many folders which are not expected to be scanned, especially folders like node_modules which can contain quadrillion of files

## Description of the Preferred Solution

Whenever we are going to scan file or dig into directory we should traverse directories tree to see if there is gitignore somewhere and check if that file or directory isn't there

Something like this one (should be optimized, but describes the idea):

```go
package main

import (
	"fmt"
	"log"
	"os"
	"path/filepath"

	ignore "github.com/sabhiram/go-gitignore"
)

func main() {
	// recursive iteration over `os.Args[1]` path
	filepath.Walk(os.Args[1], func(path string, info os.FileInfo, err error) error {
		if err != nil {
			log.Println(err)
		}

		// path == "/Users/mac/Desktop/demo/obj/Debug/net6.0/logging.AssemblyInfo.cs"
		// info.Name() == "logging.AssemblyInfo.cs"

		// inner, walk up till root or found gitignore
		var gitignore *ignore.GitIgnore
		parent := path
		for {
			gitignore, _ = ignore.CompileIgnoreFile(filepath.Join(parent, ".gitignore"))
			if gitignore != nil {
				// gitignore found, breaking loop
				break
			}
			if parent == filepath.Dir(parent) {
				// we are at root and there is no more parents, breaking loop
				break
			} else {
				parent = filepath.Dir(parent)
			}
		}

		ignored := false
		if gitignore != nil {
			ignored = gitignore.MatchesPath(path)
		}
		fmt.Printf("%v %s\n", ignored, path)
		return nil
	})
}
```

## Additional Context

<img width="522" alt="image" src="https://user-images.githubusercontent.com/88868/197689626-28771356-97a7-46b1-884d-6f1b2449577a.png">

### References

- [go-gitignore](https://github.com/sabhiram/go-gitignore) - library to parse gitignore

**Comment by gukoff** [comment]
@dustin-decker @mac2000 

How does this feature work now?

Should one explicitly pass the `.gitignore` file to the `--exclude-paths` option?
If yes:
- Are they fully compatible?
- Can one pas multiple `.gitignore` files? What if they are in the nested folders in the repository (and their paths are relative to this nested folder)?

**Comment by mac2000** [comment]
@gukoff take a closer look at #1033 pull request

at the very end issue was solved without introducing new concept of gitignore files and instead by 

bringing missing exclude paths option to git scan

plus small fix for empty lines that prevent's skipping files with space in name

so technically in the first place it was the fix for #536 and by consequence it also solved this issue as well

hope that helps
